// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "google/protobuf/timestamp.proto";

option java_multiple_files = true;
option java_package = "com.redhat.insights.inventory";
option java_outer_classname = "HostInventoryProto";
option objc_class_prefix = "HBI";

package hostinventory;

// Interface exported by the server.
service HostInventory {
  rpc CreateOrUpdate(HostList) returns (HostList) {}

  rpc Get(FilterList) returns (HostList) {}

  rpc Delete(HostList) returns (HostList) {}
}

message CanonicalFact {
    string key = 1;
    string value = 2;
}

message Fact {
    string namespace = 1;
    string key = 2;
    string value = 3;
}

message Host {
    string id = 1;
    string display_name = 2;
    map<string, string> facts = 3;
    map<string, string> canonical_facts = 4;
    map<string, string> tags = 5;
    google.protobuf.Timestamp created = 6;
    google.protobuf.Timestamp updated = 7;
    string account_number = 8;
}

message HostList {
    repeated Host hosts = 1;
}

message FilterList {
    repeated Filter filters = 1;
}

message Filter {
    repeated string ids = 1;
    repeated Fact facts = 2;
    repeated CanonicalFact canonical_facts = 3;
    repeated Fact tags = 4;
    repeated string account_numbers = 5;
}
